### PR TITLE
[Feature][Core][Starter] Refactor starter module

### PR DESCRIPTION
**Pull request by TyrantLucifer**
## Purpose of this pull request

1. Remove old transform api from starter module
2. Support transform v2 api
3. Get ready to upgrade the engine version
4. Unified the parameters for different engine

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  5. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  6. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

**Comment by TyrantLucifer** [comment]
Flink parameters:

image

Flink local submit:

image

Flink yarn-per-job submit:

image

Flink yarn-session submit:

image

**Comment by TyrantLucifer** [comment]
Spark parameters:

image

Spark submit on yarn, client mode:

image

Spark submit on yarn, cluster mode:

image

Spark submit on local:

image

**Comment by TyrantLucifer** [comment]
ST-Engine parameters:

image

image

ST-Engine submit on local:

image

ST-Engine submit on cluster:

image

**Comment by TyrantLucifer** [comment]
As you can see the images above, the parameters be unifed as the following shown:

common parameters:

|parameter|usage|
|-----------|-------|
|-c --config|Assign config file|
|-h --help|Show help usage|
|-n --name|Assign job name|
|-i --variable|Variable substitution|

For the deployment mode of spark and flink submission jobs, standardize on the `-e --deploy-mode` parameter,

Spark support `client` `cluster`,

Flink support `run` `run-application`

For spark, flink, st-engine submit target master, standardize the `--master` parameter

Spark support `yarn` `local[*]` `spark://host:port` `mesos://host:port` `k8s://https://host:port`

Flink support `local` `remote` `yarn-session` `yarn-per-job` `kubernetes-session` `yarn-application` `kubernetes-application`

ST-Engine support `local` `cluster`

**Comment by hailin0** [comment]
merge this pr from dev branch
https://github.com/apache/incubator-seatunnel/pull/3396
https://github.com/apache/incubator-seatunnel/pull/3409

File: seatunnel-core/seatunnel-core-starter/src/main/java/org/apache/seatunnel/core/starter/utils/ConfigBuilder.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    public static Config of(@NonNull String filePath) {
+        log.info("Loading config file from path: {}", filePath);
+        Config config = ConfigFactory
+                .parseFile(Paths.get(filePath).toFile())
+                .resolve(ConfigResolveOptions.defaults().setAllowUnresolved(true))
+                .resolveWith(ConfigFactory.systemProperties(),
+                        ConfigResolveOptions.defaults().setAllowUnresolved(true));
+        log.info("Parsed config file: {}", config.root().render(CONFIG_RENDER_OPTIONS));
+        return config;
+    }
```
Comment (TaoZex) [inline seatunnel-core/seatunnel-core-starter/src/main/java/org/apache/seatunnel/core/starter/utils/ConfigBuilder.java]:
```suggestion
    public static Config of(@NonNull String filePath) {
        Path path = Paths.get(filePath);
        return of(path);
    }
```

Comment (TyrantLucifer) [reply seatunnel-core/seatunnel-core-starter/src/main/java/org/apache/seatunnel/core/starter/utils/ConfigBuilder.java]: Good idea.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iotdb-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iotdb/IoTDBIT.java [hunk 2]
```diff
@@ -95,7 +94,7 @@ public void startUp() throws Exception {
         testDataset = generateTestDataSet();
     }
 
-    @TestTemplate
+    // temporary remote annotation while this pull request being merged https://github.com/apache/incubator-seatunnel/pull/3798
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iotdb-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iotdb/IoTDBIT.java]: revert

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iotdb-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iotdb/IoTDBIT.java [hunk 3]
```diff
@@ -57,8 +56,8 @@
 import java.util.stream.Stream;
 
 @Slf4j
-@DisabledOnContainer(value = {}, type = {EngineType.SEATUNNEL, EngineType.SPARK},
-    disabledReason = "There is a conflict of thrift version between IoTDB and Spark.Therefore.")
+@DisabledOnContainer(value = {}, type = {EngineType.SEATUNNEL, EngineType.SPARK, EngineType.FLINK},
+    disabledReason = "There is a conflict of thrift version between IoTDB and Spark.Therefore. Refactor starter module, so disabled in flink")
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iotdb-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iotdb/IoTDBIT.java:61]: revert

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iotdb-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iotdb/IoTDBIT.java (hunk 2 above)
Comment (TyrantLucifer) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iotdb-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iotdb/IoTDBIT.java]: Done

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iotdb-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iotdb/IoTDBIT.java (hunk 3 above)
Comment (TyrantLucifer) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-iotdb-e2e/src/test/java/org/apache/seatunnel/e2e/connector/iotdb/IoTDBIT.java:61]: Done

**Review by CalvinKirs — DISMISSED** [review]

File: seatunnel-common/src/main/java/org/apache/seatunnel/common/config/DeployMode.java [hunk 4]
```diff
@@ -17,34 +17,39 @@
...
 import java.util.Arrays;
 import java.util.Map;
 import java.util.Optional;
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
-@Getter
-@AllArgsConstructor
 public enum DeployMode {
+    /**
```
Comment (Hisoka-X) [inline seatunnel-common/src/main/java/org/apache/seatunnel/common/config/DeployMode.java:27]: Add Engine DepolyMode To Common, give me a feeling that there is no uniform parameter

File: seatunnel-core/seatunnel-core-starter/src/main/java/org/apache/seatunnel/core/starter/utils/ConfigBuilder.java [hunk 5]
```diff
@@ -15,57 +15,44 @@
...
 
-    private static final String PLUGIN_NAME_KEY = "plugin_name";
-    private final Path configFile;
-    private final Config config;
+    private static final ConfigRenderOptions CONFIG_RENDER_OPTIONS = ConfigRenderOptions.concise().setFormatted(true);
 
-    public ConfigBuilder(Path configFile) {
-        this.configFile = configFile;
-        this.config = load();
+    public ConfigBuilder() {
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-core-starter/src/main/java/org/apache/seatunnel/core/starter/utils/ConfigBuilder.java]:
```suggestion
    private ConfigBuilder() {
```

File: seatunnel-core/seatunnel-core-starter/src/main/java/org/apache/seatunnel/core/starter/utils/FileUtils.java [hunk 6]
```diff
@@ -42,12 +43,12 @@ private FileUtils() {
      * @param args args
      * @return path of the seatunnel config file.
      */
-    public static Path getConfigPath(AbstractCommandArgs args) {
-        checkNotNull(args, "args");
-        checkNotNull(args.getDeployMode(), "deploy mode");
+    public static Path getConfigPath(@NonNull AbstractCommandArgs args) {
         switch (args.getDeployMode()) {
+            case RUN:
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-core-starter/src/main/java/org/apache/seatunnel/core/starter/utils/FileUtils.java:48]: Same as above

**Review by Hisoka-X — APPROVED** [review]

**Review by CalvinKirs — APPROVED** [review]

**Review by TaoZex — APPROVED** [review]
+1

**Pull request merged**